analysis.go

```diff
@@ -79,3 +80,46 @@ func (g AnalyzerCommentsGroups) Filter(fn CommentsFilterFn) ([]AnalyzerComments,
 
 	return result, nil
 }
+
+// Dedup filters duplicated comments
+func (g AnalyzerCommentsGroups) Dedup() AnalyzerCommentsGroups {
+	var result []AnalyzerComments
+	type Key struct {
+		File string
+		Line int32
+		Text string
+	}
+
+	uniqueCommentsMap := map[Key]*Comment{}
+	for _, group := range g {
+		var newComments []*Comment
+		var dupComments int
+		for _, comment := range group.Comments {
+			key := Key{
+				File: comment.File,
+				Line: comment.Line,
+				Text: comment.Text,
+			}
+			if _, ok := uniqueCommentsMap[key]; ok {
+				dupComments++
+				continue
+			}
+
+			uniqueCommentsMap[key] = comment
+			newComments = append(newComments, comment)
+		}
+
+		if len(newComments) > 0 {
+			result = append(result, AnalyzerComments{
+				Config:   group.Config,
+				Comments: newComments,
+			})
+		}
+
+		if dupComments > 0 {
+			log.Warningf("analyzer %s generated %d duplicated comments", group.Config.Name, dupComments)
```

Comment:
I don't understand de logic behind it was decided to deduplicate comments between analyzers:
- It most likely never happen,
- even if it happens, it should not be considered a dupe because,
  - dupes from different analyzers should have different feedbacks,
  - from the [testacase](https://github.com/src-d/lookout/pull/458/files#diff-ae0924ee5d8a959a1bbe8162f321185eR71), when the dupe from the third analyzer is logged as a dupe, it can not be reproduced without the global context of source{d} Lookout, so the maintainer of that analyzer wont be able to fix it, nor even reproduce the bug.

Reply:
Regarding the decision of deduplicating the comments *as a whole* instead of doing it *by analyzer* is discussed [here](https://github.com/src-d/lookout/pull/458#discussion_r246689092).

> from the testacase, when the dupe from the third analyzer is logged as a dupe, it can not be reproduced without the global context of source{d} Lookout, so the maintainer of that analyzer wont be able to fix it, nor even reproduce the bug.

I don't know if understood what you mean. Actually an analyzer producing a comment which is the same of another analyzer is not a bug given that they are supposed to work independently. So it's not a bug that should be addressed by the analyzer's maintainer. With this PR they're filtered out to improve the user experience. Anyway your comment made me think about these two scenarios:
1. a duplicated comment is a duplicate of *another* analyzer,
2. a duplicated comment is a duplicate of the *same* analyzer.

As said before scenario 1. is not a bug of the analyzer, on the other hand scenario 2. is (in my opinion as I don't see the purpose of doing it) a bug of the analyzer.

This warning log comment is actually printed independently from being a duplicate of *another* or the *same* analyzer. As you correctly pointed out, in case of a bug in the analyzer, its maintainer should be notified. Hence I think that we could improved this warning by splitting it in order to identify whether the duplicate is originated from scenario 1 or scenario 2.

I would use the followings:
* for scenario 1:
```
log.Infof("analyzer %s generated %d comments already generated by others, skipping.", group.Config.Name, dupComments)
```
* for scenario 2:
```
log.Warningf("analyzer %s generated %d duplicated comments", group.Config.Name, dupComments)
```

Reply:
Splitted the duplicated comments counting and corresponding log messages [here](https://github.com/src-d/lookout/pull/458/commits/ae9066a13abea6891c0d4f0a8a974f24091ff1c4).

Reply:
I agree on using different messages if all of them are deduplicated.
But I don't agree on deduplicating comments that other analyzers already posted.
When @smacker [said](https://github.com/src-d/lookout/pull/458#discussion_r246689092) that the only way to know which analyzer returned a comment is the link in the footer, for me that's a valid way to distinguish them, so the user can use that link to notify each analyzer maintainer about a bug on their analyzer.

example:
consider those two comments both in the same line `113` at same `example.go` file:
-  `This line is not properly formatted; any feedback?, open issue [here](https://github.com/src-d/style-analyzer/issues)`
- `This line is not properly formatted; any feedback?, open issue [here](https://github.com/src-d/lookout-gometalint-analyzer/issues)`

The user can report an issue in both analyzers if both comments are wrong, only if different comments from different analyzers are not deduplicated.

A different problem is that we don't differentiate both comments in the messages that source{d} Lookout post on behalf the analyzers, so the user may be confused; but it's a UX problem that should be considered in a different issue, instead of dropping the second comment because the current UI is not clear enough.

Reply:
I agree with you that there's also a UX problem and that it should be addressed. In my opinion the best thing is to _remove_ the duplicates of the same analyzer and to _merge_ the duplicates of different analyzers. By doing this we can then show something like
```This line is not properly formatted; any feedback?, open issue [here](https://analyzers.sourced.tech/feedback/?analyzers=style-analyzer,lookout-gometalint-analyzer&prurl=<url of the PR>&...other useful informations...)```

I would show a simple form similar to github's one to open an issue, that on submit automatically open the issue on each analyzer's repo. I would personally be annoyed by copy-pasting the same issue for multiple analyzers.

Regarding what we want to do **now** before thinking about the UX, I don't have a strong opinion on which approach is better. I slightly prefer avoid duplicating comments globally at the cost of being able to give a feedback only to one analyzer. It depends on what we want to give more importance to (being annoyed by duplicated comments or let the use give a feedback), and also on when (and if) we want to address the UX issue.

I'd also ask @carlosms and @smacker opinion.

Reply:
final decition comes from last TFM [applications/2019-01-15#Line80](https://github.com/src-d/minutes/pull/430/files#diff-76b4489f9540c5a62d9393ef188c40bbR80)
>_To keep the comments (from different analyzers) separated. Add the name of the analyzer in a different PR._